### PR TITLE
[BugFix][Model] Recompute DeepSeek V4 hash routing weights

### PR DESCRIPTION
**Pull request by QwertyJack**
### What this PR does / why we need it?

DeepSeek V4 W4A8 can hit NaN router weights in the `sqrtsoftplus + tid2eid/hash routing` path. The hash-routing custom op still returns finite `topk_ids`, but its returned `topk_weights` can become NaN and then propagate into MoE output for long generations.

This PR keeps using `moe_gating_top_k_hash` for selected expert ids, but recomputes the corresponding weights from `router_logits` in FP32:

```python
sqrt(softplus(router_logits)).gather(topk_ids)
```

The recomputed weights are then renormalized and scaled exactly like the existing DeepSeek V4 path. The W4A8 MoE path also keeps `topk_weights` in FP32 before expert scaling to avoid losing precision before MC2/ALLTOALL combine can promote them again.

The branch was created from `main@b26970e597222029d624261a8738a40f2d33ad17`.

### Does this PR introduce _any_ user-facing change?

No. This is an internal accuracy fix for DeepSeek V4 W4A8 MoE routing on Ascend.

### How was this patch tested?

- `python -m py_compile vllm_ascend/ops/fused_moe/experts_selector.py vllm_ascend/quantization/methods/w4a8.py tests/ut/ops/test_moe_hash_routing_input_ids.py`
- `ruff check vllm_ascend/ops/fused_moe/experts_selector.py vllm_ascend/quantization/methods/w4a8.py tests/ut/ops/test_moe_hash_routing_input_ids.py`
- `pytest -q tests/ut/ops/test_moe_hash_routing_input_ids.py tests/ut/quantization/methods/test_w4a8.py::TestAscendW4A8DynamicFusedMoEMethod::test_apply_comprehensive`

The new unit tests cover comm-aligned input ids for hash routing and the regression case where the custom op returns NaN weights while `topk_ids` and `router_logits` remain finite.

Manual validation was also run on DeepSeek-V4-Pro W4A8 with TP16DP1EP, eager mode, prefix cache disabled, FlashComm/FUSED_MC2 disabled, `HCCL_DETERMINISTIC=true`, and `HCCL_BUFFSIZE=512`. A long deterministic completion that previously produced abnormal tokens no longer propagated non-finite values through the effective MoE/hidden/logits path after this fix.

- vLLM version: v0.20.2
- vLLM main: https://github.com/vllm-project/vllm/commit/39910f2b25aacc09f5e7f166cdf0030b19f8b9e8

### CODE REVIEW

**Comment by QwertyJack** [comment]
fixed by PR #9692
